frame/support/procedural/src/benchmark.rs

```diff
@@ -0,0 +1,816 @@
+struct ParamDef {
+	name: String,
+	typ: Type,
+	start: u32,
+	end: u32,
+}
+
+/// Allows easy parsing of the `<10, 20>` component of `x: Linear<10, 20>`.
+#[derive(Parse)]
+struct RangeArgs {
+	_lt_token: Lt,
+	start: LitInt,
+	_comma: Comma,
+	end: LitInt,
+	_gt_token: Gt,
+}
+
+#[derive(Clone, Debug)]
+struct BenchmarkAttrs {
+	skip_meta: bool,
+	extra: bool,
+}
+
+/// Represents a single benchmark option
+enum BenchmarkAttrKeyword {
+	Extra,
+	SkipMeta,
+}
+
+impl syn::parse::Parse for BenchmarkAttrKeyword {
+	fn parse(input: ParseStream) -> Result<Self> {
+		let lookahead = input.lookahead1();
+		if lookahead.peek(keywords::extra) {
+			let _extra: keywords::extra = input.parse()?;
+			return Ok(BenchmarkAttrKeyword::Extra)
+		} else if lookahead.peek(keywords::skip_meta) {
+			let _skip_meta: keywords::skip_meta = input.parse()?;
+			return Ok(BenchmarkAttrKeyword::SkipMeta)
+		} else {
+			return Err(lookahead.error())
+		}
+	}
+}
+
+impl syn::parse::Parse for BenchmarkAttrs {
+	fn parse(input: ParseStream) -> syn::Result<Self> {
+		let lookahead = input.lookahead1();
+		if !lookahead.peek(Paren) {
+			let _nothing: Nothing = input.parse()?;
+			return Ok(BenchmarkAttrs { skip_meta: false, extra: false })
+		}
+		let content;
+		let _paren: Paren = parenthesized!(content in input);
+		let mut extra = false;
+		let mut skip_meta = false;
+		let args = Punctuated::<BenchmarkAttrKeyword, Token![,]>::parse_terminated(&content)?;
+		for arg in args.into_iter() {
+			match arg {
+				BenchmarkAttrKeyword::Extra => {
+					if extra {
+						return Err(content.error("`extra` can only be specified once"))
+					}
+					extra = true;
+				},
+				BenchmarkAttrKeyword::SkipMeta => {
+					if skip_meta {
+						return Err(content.error("`skip_meta` can only be specified once"))
+					}
+					skip_meta = true;
+				},
+			}
+		}
+		Ok(BenchmarkAttrs { extra, skip_meta })
+	}
+}
+
+/// Represents the parsed extrinsic call for a benchmark
+#[derive(Clone)]
+enum ExtrinsicCallDef {
+	Encoded { origin: Expr, expr_call: ExprCall }, // no block, just an extrinsic call
+	Block(ExprBlock),                              // call is somewhere in the block
+}
+
+/// Represents a parsed `#[benchmark]` or `#[instance_banchmark]` item.
+#[derive(Clone)]
+struct BenchmarkDef {
+	params: Vec<ParamDef>,
+	setup_stmts: Vec<Stmt>,
+	extrinsic_call: ExtrinsicCallDef,
+	verify_stmts: Vec<Stmt>,
+	extra: bool,
+	skip_meta: bool,
+}
+
+impl BenchmarkDef {
+	/// Constructs a [`BenchmarkDef`] by traversing an existing [`ItemFn`] node.
+	pub fn from(item_fn: &ItemFn, extra: bool, skip_meta: bool) -> Result<BenchmarkDef> {
+		let mut i = 0; // index of child
+		let mut params: Vec<ParamDef> = Vec::new();
+
+		// parse params such as "x: Linear<0, 1>"
+		for arg in &item_fn.sig.inputs {
+			let mut name: Option<String> = None;
+			let mut typ: Option<&Type> = None;
+			let mut start: Option<u32> = None;
+			let mut end: Option<u32> = None;
+			if let FnArg::Typed(arg) = arg {
```

Comment:
This entire code block here is fantastic for using let-else instead of if-let. You can eliminate pretty much all of the `Option` let bindings as well.